packages/api/src/clickhouse/index.ts

```diff
@@ -663,78 +684,100 @@ export const getMetricsChart = async ({
-          ? `AVG(value) as data`
-          : aggFn === AggFn.Max
-          ? `MAX(value) as data`
-          : aggFn === AggFn.Min
-          ? `MIN(value) as data`
-          : `quantile(${
-              aggFn === AggFn.P50
-                ? '0.5'
-                : aggFn === AggFn.P90
-                ? '0.90'
-                : aggFn === AggFn.P95
-                ? '0.95'
-                : '0.99'
-            })(value) as data`,
-      );
-      break;
-    case 'Sum':
-      selectClause.push(
-        aggFn === AggFn.Count
-          ? 'COUNT(delta) as data'
-          : aggFn === AggFn.Sum
-          ? `SUM(delta) as data`
-          : aggFn === AggFn.Avg
-          ? `AVG(delta) as data`
-          : aggFn === AggFn.Max
-          ? `MAX(delta) as data`
-          : aggFn === AggFn.Min
-          ? `MIN(delta) as data`
-          : `quantile(${
-              aggFn === AggFn.P50
-                ? '0.5'
-                : aggFn === AggFn.P90
-                ? '0.90'
-                : aggFn === AggFn.P95
-                ? '0.95'
-                : '0.99'
-            })(delta) as data`,
-      );
-      break;
-    default:
-      logger.error(`Unsupported data type: ${dataType}`);
-      break;
+  const isRate = isRateAggFn(aggFn);
+
+  if (dataType === 'Gauge' || dataType === 'Sum') {
+    selectClause.push(
+      aggFn === AggFn.Count
+        ? 'COUNT(value) as data'
+        : aggFn === AggFn.Sum
+        ? `SUM(value) as data`
+        : aggFn === AggFn.Avg
+        ? `AVG(value) as data`
+        : aggFn === AggFn.Max
+        ? `MAX(value) as data`
+        : aggFn === AggFn.Min
+        ? `MIN(value) as data`
+        : aggFn === AggFn.SumRate
+        ? `SUM(rate) as data`
+        : aggFn === AggFn.AvgRate
+        ? `AVG(rate) as data`
+        : aggFn === AggFn.MaxRate
+        ? `MAX(rate) as data`
+        : aggFn === AggFn.MinRate
+        ? `MIN(rate) as data`
+        : `quantile(${
+            aggFn === AggFn.P50 || aggFn === AggFn.P50Rate
+              ? '0.5'
+              : aggFn === AggFn.P90 || aggFn === AggFn.P90Rate
+              ? '0.90'
+              : aggFn === AggFn.P95 || aggFn === AggFn.P95Rate
+              ? '0.95'
+              : '0.99'
+          })(${isRate ? 'rate' : 'value'}) as data`,
+    );
+  } else {
+    logger.error(`Unsupported data type: ${dataType}`);
   }
 
+  const rateMetricSource = SqlString.format(
+    `
+SELECT
+  if(
+    runningDifference(value) < 0
+    OR neighbor(_string_attributes, -1, _string_attributes) != _string_attributes,
+    nan,
+    runningDifference(value)
+  ) AS rate,
+  ts_bucket as timestamp,
+  _string_attributes,
+  min_name as name
+FROM
+  (
+    SELECT
+      toStartOfInterval(timestamp, INTERVAL ?) as ts_bucket,
+      min(value) as value,
+      _string_attributes,
+      min(name) as min_name
+    FROM
+      ??
+    WHERE
+      name = ?
+      AND data_type = ?
+      AND (?)
+    GROUP BY
+      _string_attributes,
```

Comment:
can we also group by the `name` here ?

Reply:
Since `name` is fixed, it didn't seem necessary so I just did `min(name)`, though now I wonder if `group by name` is actually more performant than `min(name)`. I think right now the query is correct - but I realize that there's possibly another option where we just inject the `name` as a fixed value in the query itself so Clickhouse doesn't have to do any additional processing on the name column as well.

On the perf side - I'm realizing that the indexing order has an effect here on the grouping as well (and it seems like a load of other things), and we can probably come back and do some benchmarking/see if there's an optimization to be done here later as CH's grouping optimization logic is quite complex.